### PR TITLE
APPS-624 build globalworkflow from workflow

### PR DESCRIPTION
**Pull request by YuxinShi0423**
This PR is to support building global workflow from workflow (both wdl and non-wdl ones).

Requirements included:

- Detect that local workflow is built with dxCompiler or dxWDL of an older version. If so, refuse to create a global workflow
- Notify the global workflow builder to check workflow description for the dependencies and share them with authorized users when detecting a dxCompiler-build local workflow with dependencies. It should work for non-WDL workflows as well
- Global workflow description should be cloned from the local workflow description, including the dependencies listing generated by the dxCompiler

### CODE REVIEW

File: src/python/dxpy/workflow_builder.py
Comment (r-i-v-a) [inline src/python/dxpy/workflow_builder.py:97]: Eventually edit message to dxCompiler >= version when we know what version

File: src/python/dxpy/workflow_builder.py
Comment (r-i-v-a) [inline src/python/dxpy/workflow_builder.py:416]: This same thing is getting checked multiple times, for each executable?

File: src/python/dxpy/workflow_builder.py
Comment (r-i-v-a) [inline src/python/dxpy/workflow_builder.py:93]: You don't want to always enforce it's the latest version of dxCompiler -- that is going to change over time. The version to compare against will be 1 specific version (the release of dxCompiler where all features on that side are implemented.) Maybe set a placeholder number here with a comment to update it later, like 100.100.100

File: src/python/dxpy/workflow_builder.py
Comment (r-i-v-a) [reply src/python/dxpy/workflow_builder.py:416]: I guess it will stop on the first applet if the exception will be thrown. Probably could be more nicely written as if len(workflow_enabled_regions) > 1 and filter(check it startsWith applet-) returns a non-empty result, then throw the exception. If only 1 region, the left side of the "and" will be false, and no need to proceed to the more computationally expensive right side of the "and."

File: src/python/dxpy/workflow_builder.py
Comment (r-i-v-a) [inline src/python/dxpy/workflow_builder.py:578]: Have to look in more detail at what happens inside that global_workflow_new() in order to review this.

File: src/python/dxpy/workflow_builder.py
Comment (YuxinShi0423) [reply src/python/dxpy/workflow_builder.py:416]: I changed it because it should be allowed to assign a single region where the applet is. But the old code is throwing an error.

Comment (YuxinShi0423) [reply src/python/dxpy/workflow_builder.py:416]: But if we request one region that does not contain the applet, checking the `len(workflow_enabled_regions)` could not detect the conflict.

File: src/python/dxpy/workflow_builder.py
Comment (YuxinShi0423) [inline src/python/dxpy/workflow_builder.py:456]: The enabled region(s) will be the union of the following:
- regionalOptions of source workflow spec(if exists)
- `--region` from CLI (if given)
- the permitted regions of billTo
- the dependent executables listed in the source workflow's description.

File: src/python/dxpy/workflow_builder.py
Comment (YuxinShi0423) [inline src/python/dxpy/workflow_builder.py:437]: I changed it to project context id based on http://autodoc.dnanexus.com/bindings/python/current/dxpy.html since we are not running a job when building gwf. @commandlinegirl

File: src/python/dxpy/workflow_builder.py
Comment (YuxinShi0423) [reply src/python/dxpy/workflow_builder.py:382]: for example if the `workflow_enabled_regions` is `azure:us-west` but the `applet-xxxx` is in `aws:us-west`, the previous code will not detect it.

File: src/python/dxpy/workflow_builder.py
Comment (mhrvol) [inline src/python/dxpy/workflow_builder.py:274]: For apps we use `dxapp.json` ( in src/python/dxpy/scripts/dx_build_app.py). Maybe we should change that as well, just to stay consistent?

File: src/python/dxpy/workflow_builder.py
Comment (mhrvol) [inline src/python/dxpy/workflow_builder.py:372]:
```suggestion
    If the workflow contains any applets, then the workflow can be currently enabled
```

File: src/python/dxpy/executable_builder.py
Comment (mhrvol) [inline src/python/dxpy/executable_builder.py:127]: Where is this function used?

File: src/python/dxpy/executable_builder.py
Comment (mhrvol) [inline src/python/dxpy/executable_builder.py:132]: it looks like a function to add to DXExecutable etc.

File: src/python/dxpy/scripts/dx.py
Comment (mhrvol) [inline src/python/dxpy/scripts/dx.py:2544]: Can we remove this comment?

File: src/python/dxpy/executable_builder.py
Comment (mhrvol) [inline src/python/dxpy/executable_builder.py:168]: Do we want to add developer or remove this todo??

File: src/python/dxpy/workflow_builder.py
Comment (r-i-v-a) [inline src/python/dxpy/workflow_builder.py:91]: Before merging, this should be updated to the actual dxCompiler version where the feature is supported (which we don't know yet until dxCompiler part is done).

File: src/python/dxpy/workflow_builder.py
Comment (r-i-v-a) [inline src/python/dxpy/workflow_builder.py:94]:
```suggestion
            raise WorkflowBuilderException("Source workflow {} is not compiled using dxCompiler (version>={}) that supports creating global workflows.".format(json_spec["name"], supported_version))
```

**Review by r-i-v-a — CHANGES_REQUESTED** [review]
Minor comments about doc strings. Otherwise, looks good, pending final dxCompiler supported_version to use. @YuxinShi0423

File: src/python/dxpy/executable_builder.py
Comment (YuxinShi0423) [reply src/python/dxpy/executable_builder.py]: Deleted. Thank you :D

File: src/python/dxpy/workflow_builder.py
Comment (YuxinShi0423) [inline src/python/dxpy/workflow_builder.py:110]: @r-i-v-a @emiloslavsky as said here:https://confluence.internal.dnanexus.com/display/TTN/Global+WDL+workflows

> If we detect a dxCompiler-built local workflow with dependencies and --brief is not specified, notify the global workflow builder to check workflow description for these dependencies and share them with authorized users

So shall I just put a short message like above, or parse the description and print the whole list out?

Comment (YuxinShi0423) [reply src/python/dxpy/workflow_builder.py:110]: > Updated "details" (if static instance type) looks like
>     "details": {
>         "staticInstanceTypeSelection": true,
>     },
> If dynamic instance type selection, that field will not be in "details."

File: src/python/dxpy/workflow_builder.py
Comment (YuxinShi0423) [inline src/python/dxpy/workflow_builder.py:103]: > Updated "details" (if static instance type) looks like
> ```
>      "details": {
>          "staticInstanceTypeSelection": true,
>      },
> ```
> If dynamic instance type selection, that field will not be in "details."
@r-i-v-a @commandlinegirl

File: src/python/dxpy/workflow_builder.py
Comment (YuxinShi0423) [inline src/python/dxpy/workflow_builder.py:646]: Here checking instance type selection and unbundled dependency only run when --brief is false, because it is just sending notifications to the user without changing the workflow.

File: src/python/dxpy/workflow_builder.py
Comment (r-i-v-a) [reply src/python/dxpy/workflow_builder.py]: Just put a short message that tells them to check the metadata details field. @YuxinShi0423

File: src/python/dxpy/workflow_builder.py
Comment (r-i-v-a) [reply src/python/dxpy/workflow_builder.py:286]: Yes, I think @mhrvol means the error message is worded that way when building an app from an applet. @YuxinShi0423

File: src/python/dxpy/workflow_builder.py
Comment (r-i-v-a) [inline src/python/dxpy/workflow_builder.py:702]: Shouldn't it be `".format()` without a space?

**Comment by YuxinShi0423** [comment]
Built globalworkflow-G6XZ1bj018Zyx444749fY0bz dependency_report_wf1-0.0.3.json from workflow-G6XY6500yzZyx444749fY0ZK
https://staging.dnanexus.com/projects/Fy9QqgQ0yzZbg9KXKP4Jz6Yq/monitor/analysis/G6XZ20j0yzZxB2597z1f0zP3

@r-i-v-a

**Review by r-i-v-a — APPROVED** [review]

File: src/python/dxpy/executable_builder.py
Comment (YuxinShi0423) [inline src/python/dxpy/executable_builder.py:189]: I reverted my previous chanage, which only check if --region is a subset of regionalOptions from the json spec. It sounds more consistent with the helpstring `help="Enable the app/globalworkflow in this region. This flag can be specified multiple times to enable the app/globalworkflow in multiple regions. If --region is not specified, then the enabled region(s) will be determined by 'regionalOptions' in dxapp.json, or the project context."` , but I don't want to change the behavior here since I am not sure about the side effect when building app/gwf.

File: src/python/dxpy/workflow_builder.py
Comment (YuxinShi0423) [reply src/python/dxpy/workflow_builder.py:286]: I have checked `src/python/dxpy/scripts/dx_build_app.py` and all messages containing "dxapp.json" are inside the building process of apps from dxapp.json.

File: src/python/dxpy/app_builder.py
Comment (YuxinShi0423) [inline src/python/dxpy/app_builder.py:907]: @mhrvol @r-i-v-a the messages in app_builder are updated. Thanks for the suggestion!

File: src/python/dxpy/workflow_builder.py
Comment (YuxinShi0423) [reply src/python/dxpy/workflow_builder.py:286]: But in `app_builder.py` there are some messages that need to be updated. Please check the other comment.

File: src/python/dxpy/workflow_builder.py
Comment (commandlinegirl) [inline src/python/dxpy/workflow_builder.py:395]: How about using dx bindings here:
```suggestion
            applet_project = dxpy.DXApplet(exect).project
            applet_region = dxpy.DXProject(applet_project).region
```

File: src/python/dxpy/workflow_builder.py
Comment (commandlinegirl) [inline src/python/dxpy/workflow_builder.py:397]: After this we can add a `break` since it's all the applets will be in the same region so no need to iterate over the remaining ones. This pertains to the workflow used in the --from option as well as a nested workflow.

File: src/python/dxpy/workflow_builder.py
Comment (commandlinegirl) [inline src/python/dxpy/workflow_builder.py:386]: Please update this docstring to say what it returns now.

**Review by commandlinegirl — CHANGES_REQUESTED** [review]

File: src/python/test/test_dxclient.py
Comment (commandlinegirl) [inline src/python/test/test_dxclient.py:6979]: All these tests are negative tests. The positive tests will be in the dxCompiler (so as not to add the dxCompiler dependency to tests in dx-toolkit)?

Comment (r-i-v-a) [reply src/python/test/test_dxclient.py:6979]: Yes, I will put the positive tests in dxCompiler. APPS-627

File: src/python/dxpy/app_builder.py
Comment (r-i-v-a) [reply src/python/dxpy/app_builder.py:907]: I thought @mhrvol 's suggestion was to go the other way, and use dxapp.json / dxworkflow.json throughout? dxapp.json is used often enough in our user-facing documentation that it should be clear to the user what it refers to https://documentation.dnanexus.com/developer/apps/app-metadata , maybe clearer than calling it "app spec."

File: src/python/dxpy/executable_builder.py
Comment (r-i-v-a) [reply src/python/dxpy/executable_builder.py:190]: I don't see a problem with leaving the original behavior here. The way I read it, you are meant to either use `--region` without restrictions if the JSON doesn't contain `regionalOptions`, or not use the flag if `regionalOptions` are present. In case both are used at the same time, it enforces that they match exactly to avoid ambiguity.

File: src/python/dxpy/app_builder.py
Comment (r-i-v-a) [reply src/python/dxpy/app_builder.py:907]: Ok, I defer to @commandlinegirl 's comment on the other thread, so it can stay as "app / workflow spec."

File: src/python/test/test_dxclient.py
Comment (YuxinShi0423) [reply src/python/test/test_dxclient.py:6967]: Added some comments about testing wdl wf

File: src/python/dxpy/workflow_builder.py
Comment (YuxinShi0423) [inline src/python/dxpy/workflow_builder.py:89]: @commandlinegirl the version number is moved to the top of the function.

File: src/python/dxpy/workflow_builder.py
Comment (YuxinShi0423) [inline src/python/dxpy/workflow_builder.py:386]: The docstring is updated @commandlinegirl

File: src/python/dxpy/workflow_builder.py
Comment (commandlinegirl) [reply src/python/dxpy/workflow_builder.py:89]: Thanks

File: src/python/dxpy/executable_builder.py
Comment (commandlinegirl) [reply src/python/dxpy/executable_builder.py:190]: After some consideration and discussions with Yuxin (and her investigation), we've decided it's best to tackle `--region` separately and its behavior won't be changed here.

**Review by commandlinegirl — APPROVED** [review]
LGTM, thank you!

**Pull request merged**
